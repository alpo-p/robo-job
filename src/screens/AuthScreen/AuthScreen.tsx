import React, { useEffect } from 'react'
import { Text } from 'react-native'
import * as Google from 'expo-google-app-auth'
import { IOS_CLIENT_ID } from '@env'
import {
  getAuth,
  GoogleAuthProvider,
  signInWithCredential,
} from '@firebase/auth'
import { useNavigation } from '@react-navigation/native'
import { SignInWithGoogleButton } from './components/SignInWithGoogleButton'

import AuthStorage from '../../utils/authStorage'
import useAuthStorage from '../../hooks/useAuthStorage'
import { SafeContainer } from '../../common/components/SafeContainer'

const AuthScreen: React.FC = () => {
  const auth = getAuth()
  const authStorage: AuthStorage = useAuthStorage()
  const navigation = useNavigation()

  const signInWithTokenAndNavigateToRoot = async (token: string) => {
    const credential = GoogleAuthProvider.credential(null, token)
    await signInWithCredential(auth, credential)
    // @ts-ignore
    navigation.navigate('RootNavigator')
  }

  useEffect(() => {
    const tryToSignInUserFromAuthStorage = async () => {
      const token = await authStorage.getAccessToken()
      if (token) {
        signInWithTokenAndNavigateToRoot(token)
      }
    }
    tryToSignInUserFromAuthStorage()
  })

  const signInWithGoogleAndReturnAccessToken = async () => {
    try {
      const result = await Google.logInAsync({
        iosClientId: IOS_CLIENT_ID,
        scopes: ['email'],
      })

      if (result.type === 'success') {
        return result.accessToken
      }
    } catch (e) {
      console.error('Error signing in with Google')
    }
    return null
  }

  const handleSignIn = async () => {
    const token = await signInWithGoogleAndReturnAccessToken()
    if (token) {
      await authStorage.setAccessToken(token)
      signInWithTokenAndNavigateToRoot(token)
    }
  }

  return (
    <SafeContainer>
      <Text>Sign in by clicking below</Text>
      <SignInWithGoogleButton onPress={() => handleSignIn()} />
    </SafeContainer>
  )
}

export default AuthScreen
