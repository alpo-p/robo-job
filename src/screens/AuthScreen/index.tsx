import React from 'react'
import { Text } from 'react-native'
import * as Google from 'expo-google-app-auth'
import { IOS_CLIENT_ID } from '@env'
import {
  getAuth,
  GoogleAuthProvider,
  signInWithCredential,
} from '@firebase/auth'
import { useNavigation } from '@react-navigation/native'
import { SignInWithGoogleButton } from './components/SignInWithGoogleButton'

import AuthStorage from '../../utils/authStorage'
import useAuthStorage from '../../hooks/useAuthStorage'
import { SafeContainer } from '../../common/components/SafeContainer'
import { RoboButton } from '../../common/components/RoboButton'
import { NavigationPropType } from '../../Navigator'

const AuthScreen: React.FC = () => {
  const auth = getAuth()
  const authStorage: AuthStorage = useAuthStorage()
  const navigation = useNavigation<NavigationPropType>()

  const signInWithTokenAndNavigateToRoot = async (token: string) => {
    const credential = GoogleAuthProvider.credential(null, token)

    // TODO: This throws an error Firebase: Unsuccessful check authorization response from Google: {]
    // when have waited for a while an come back
    // The token is still the same, but for some reason it doens't work
    // maybe Google.logInAsync() is needed for some reason again..?
    // https://stackoverflow.com/questions/40450850/error-google-id-token-is-not-allowed-to-be-used-with-this-application-firebase
    await signInWithCredential(auth, credential)
    // @ts-ignore
    navigation.navigate('RootNavigator')
  }

  /*
  * This is used to enable autochecking of "is the user logged in?"
  useEffect(() => {
    const tryToSignInUserFromAuthStorage = async () => {
      const token = await authStorage.getAccessToken()
      if (token) {
        await signInWithTokenAndNavigateToRoot(token)
      }
    }

    tryToSignInUserFromAuthStorage()
  })
  */

  const signInWithGoogleAndReturnAccessToken = async () => {
    try {
      const result = await Google.logInAsync({
        iosClientId: IOS_CLIENT_ID,
        scopes: ['email'],
      })

      if (result.type === 'success') {
        return result.accessToken
      }
    } catch (e) {
      console.error('Error signing in with Google')
    }
    return null
  }

  const handleSignIn = async () => {
    const token = await signInWithGoogleAndReturnAccessToken()
    if (token) {
      await authStorage.setAccessToken(token)
      await signInWithTokenAndNavigateToRoot(token)
    }
  }

  const signInAsDemo = () => navigation.navigate('RootNavigator')

  return (
    <SafeContainer>
      <Text>Sign in by clicking below</Text>
      <SignInWithGoogleButton onPress={() => handleSignIn()} />
      <RoboButton onPress={() => signInAsDemo()} title="Sign in as demo" />
    </SafeContainer>
  )
}

export default AuthScreen
