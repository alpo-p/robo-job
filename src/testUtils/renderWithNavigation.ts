/* eslint-disable import/no-extraneous-dependencies */
import { JSXElementConstructor, ReactElement } from 'react'
import renderer from 'react-test-renderer'

type PropType = ReactElement<any, string | JSXElementConstructor<any>>

const mockedNavigate = jest.fn()

const mockedTheme = {
  colors: {
    primary: '#000000',
    background: '#000000',
    card: '#000000',
    text: '#000000',
    border: '#000000',
    notification: '#000000',
  },
}

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native')
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
    useTheme: () => mockedTheme,
    useColorScheme: () => 'dark',
  }
})

const renderWithNavigation = (children: PropType) => {
  return renderer.create(children)
}

export default renderWithNavigation
