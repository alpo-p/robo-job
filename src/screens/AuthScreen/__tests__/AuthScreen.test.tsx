import React from 'react'
import renderWithNavigation from '../../../testUtils/renderWithNavigation'
import AuthScreen from '../index'

describe('AuthScreen', () => {
  it('should match snapshot', () => {
    const tree = renderWithNavigation(<AuthScreen />)
    expect(tree).toMatchSnapshot()
  })
})
