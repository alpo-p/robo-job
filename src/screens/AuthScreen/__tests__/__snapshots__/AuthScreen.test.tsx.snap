// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthScreen should match snapshot 1`] = `
Array [
  <RCTScrollView
    horizontal={true}
    onScroll={[Function]}
    pagingEnabled={true}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <RNCSafeAreaView
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://cdn3.iconfinder.com/data/icons/movies-solid/60/54_-_Robocop_video_film_cinema_movie-1024.png",
              }
            }
            style={
              Object {
                "height": 100,
                "tintColor": undefined,
                "width": 100,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "black",
                "fontFamily": "DamascusBold",
                "fontSize": 40,
                "textTransform": "none",
              }
            }
          >
            robo job
          </Text>
        </View>
        <View
          style={
            Object {
              "marginTop": 8,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "DamascusLight",
              "fontSize": 20,
              "marginTop": -8,
              "textTransform": "none",
            }
          }
        >
          The easiest way to find a job
        </Text>
        <View
          style={
            Object {
              "bottom": 80,
              "position": "absolute",
              "width": "90%",
            }
          }
        />
      </RNCSafeAreaView>
      <RNCSafeAreaView
        style={
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "justifyContent": "center",
            "padding": 40,
            "width": 750,
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "DamascusBold",
              "fontSize": 20,
              "textTransform": "none",
            }
          }
        >
          Search for jobs
        </Text>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "DamascusLight",
              "fontSize": 16,
              "marginTop": 0,
              "textAlign": "center",
              "textTransform": "none",
            }
          }
        >
          ...and then swipe through them
        </Text>
      </RNCSafeAreaView>
      <RNCSafeAreaView
        style={
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "justifyContent": "center",
            "padding": 40,
            "width": 750,
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "DamascusBold",
              "fontSize": 20,
              "textTransform": "none",
            }
          }
        >
          Save the jobs you like
        </Text>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "DamascusLight",
              "fontSize": 16,
              "marginTop": 0,
              "textAlign": "center",
              "textTransform": "none",
            }
          }
        >
          Your profile will not be shared when liking
        </Text>
      </RNCSafeAreaView>
      <RNCSafeAreaView
        style={
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "justifyContent": "center",
            "padding": 40,
            "width": 750,
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "DamascusBold",
              "fontSize": 20,
              "textTransform": "none",
            }
          }
        >
          Quick apply
        </Text>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontFamily": "DamascusLight",
              "fontSize": 16,
              "marginTop": 0,
              "textAlign": "center",
              "textTransform": "none",
            }
          }
        >
          Apply to jobs by answering a couple of questions by the company
        </Text>
      </RNCSafeAreaView>
      <RNCSafeAreaView
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://cdn3.iconfinder.com/data/icons/movies-solid/60/54_-_Robocop_video_film_cinema_movie-1024.png",
              }
            }
            style={
              Object {
                "height": 100,
                "tintColor": undefined,
                "width": 100,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "black",
                "fontFamily": "DamascusBold",
                "fontSize": 40,
                "textTransform": "none",
              }
            }
          >
            robo job
          </Text>
        </View>
        <View
          style={
            Object {
              "marginTop": 8,
            }
          }
        />
        <View
          style={
            Object {
              "bottom": 80,
              "position": "absolute",
              "width": "90%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#000000",
                "borderRadius": 8,
                "opacity": 1,
                "padding": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 2,
                  }
                }
              >
                <Text />
              </View>
              <Text
                style={
                  Object {
                    "color": "white",
                    "flex": 5,
                    "fontFamily": "DamascusBold",
                    "fontSize": 16,
                    "textTransform": "none",
                  }
                }
              >
                Sign in with phone
              </Text>
            </View>
          </View>
        </View>
      </RNCSafeAreaView>
    </View>
  </RCTScrollView>,
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 40,
        "flexDirection": "row",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 5,
          "height": 10,
          "marginLeft": 10,
          "opacity": 1,
          "width": 10,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 5,
          "height": 10,
          "marginLeft": 10,
          "opacity": 0.2,
          "width": 10,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 5,
          "height": 10,
          "marginLeft": 10,
          "opacity": 0.2,
          "width": 10,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 5,
          "height": 10,
          "marginLeft": 10,
          "opacity": 0.2,
          "width": 10,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 5,
          "height": 10,
          "marginLeft": 10,
          "opacity": 0.2,
          "width": 10,
        }
      }
    />
  </View>,
]
`;
